neuro_san/service/grpc_agent_service.py

```diff
@@ -0,0 +1,150 @@
+
+import json
+import grpc
+
+from google.protobuf.json_format import MessageToDict
+from google.protobuf.json_format import Parse
+
+from leaf_server_common.server.grpc_metadata_forwarder import GrpcMetadataForwarder
+from leaf_server_common.server.request_logger import RequestLogger
+
+from neuro_san.api.grpc import agent_pb2 as service_messages
+from neuro_san.api.grpc import agent_pb2_grpc
+from neuro_san.internals.interfaces.agent_tool_factory_provider import AgentToolFactoryProvider
+from neuro_san.service.agent_server_logging import AgentServerLogging
+from neuro_san.service.agent_service import AgentService
+
+
+class GrpcAgentService(agent_pb2_grpc.AgentServiceServicer):
+    """
+    A gRPC implementation of the Neuro-San Agent Service.
+    """
+
+    # pylint: disable=too-many-arguments,too-many-positional-arguments
+    def __init__(self,
+                 request_logger: RequestLogger,
+                 security_cfg: Dict[str, Any],
+                 agent_name: str,
+                 tool_registry_provider: AgentToolFactoryProvider,
+                 server_logging: AgentServerLogging):
+        """
+        Set the gRPC interface up for health checking so that the service
+        will be opened to callers when the mesh sees it operational, if this
+        is not done the mesh will treat the service instance as non-functional
+
+        :param request_logger: The instance of the RequestLogger that helps
+                    keep track of stats
+        :param security_cfg: A dictionary of parameters used to
+                        secure the TLS and the authentication of the gRPC
+                        connection.  Supplying this implies use of a secure
+                        GRPC Channel.  If None, uses insecure channel.
+        :param agent_name: The agent name for the service
+        :param tool_registry_provider: The AgentToolFactoryProvider to use for the session.
+        :param server_logging: An AgentServerLogging instance initialized so that
+                        spawned asyncrhonous threads can also properly initialize
+                        their logging.
+        """
+        self.forwarder: GrpcMetadataForwarder = server_logging.get_forwarder()
+        self.service: AgentService =\
+            AgentService(request_logger,
+                         security_cfg,
+                         agent_name,
+                         tool_registry_provider,
+                         server_logging)
+
+    def get_request_count(self) -> int:
+        """
+        :return: The number of currently active requests
+        """
+        return self.service.get_request_count()
+
+    # pylint: disable=no-member
+    def Function(self, request: service_messages.FunctionRequest,
+                 context: grpc.ServicerContext) \
+            -> service_messages.FunctionResponse:
+        """
+        Allows a client to get the outward-facing function for the agent
+        served by this service.
+
```

Comment:
For all our API methods schema is the same:
convert parameters from protocol (gRPC) specific objects to Python dictionaries;
prepare metadata we will need;
delegate execution to AgentService instance;
convert resulting dictionary (or generator of dictionaries) to gRPC representation.